haystack/pipelines/utils.py

```diff
@@ -0,0 +1,264 @@
+import inspect
+import logging
+import re
+import sys
+import networkx as nx
+from typing import Any, Dict, List, Optional
+from networkx import DiGraph
+from haystack.pipelines.config import (
+    build_component_dependency_graph,
+    get_component_definitions,
+    get_pipeline_definition,
+    validate_config,
+)
+
+from haystack.schema import EvaluationResult
+
+
+logger = logging.getLogger(__name__)
+
+
+MODULE_NOT_FOUND = "MODULE_NOT_FOUND"
+CODE_GEN_ALLOWED_IMPORTS = ["haystack.document_stores", "haystack.nodes", "haystack.pipelines"]
+CAMEL_CASE_TO_SNAKE_CASE_REGEX = re.compile(r"(?<=[a-z])(?=[A-Z0-9])")
+
+
+def camel_to_snake_case(input: str) -> str:
+    return CAMEL_CASE_TO_SNAKE_CASE_REGEX.sub("_", input).lower()
+
+
+class _PipelineCodeGen:
```

Comment:
General code style note: I've never seen the point of classes with only `classmethod`s.

We can have two modules, `_pipeline_codegen` and `_pipeline_eval_report` and import them whole. Then, in `base.py`, you `import haystack.pipelines._pipeline_codegen as codegen` and then do `codegen.generate_code()`.

What the advantage of the `_PipelineCodeGen` class? It feels very Java to me :grin:

Reply:
done.